File: dataset/CoT/GEMINI_gemini-2.5-pro/Task186.py
import heapq
from typing import List, Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None

        min_heap = []
        # In Python, if two items have the same priority, the heap might try to compare
        # the next item in the tuple (the ListNode object), which is not comparable.
        # Adding a unique counter `i` as a tie-breaker avoids this issue.
        counter = 0 
        for head in lists:
            if head:
                heapq.heappush(min_heap, (head.val, counter, head))
                counter += 1

        dummy = ListNode(0)
        tail = dummy

        while min_heap:
            val, _, node = heapq.heappop(min_heap)
            tail.next = node
            tail = tail.next

            if node.next:
                heapq.heappush(min_heap, (node.next.val, counter, node.next))
                counter += 1
        
        return dummy.next

# Helper function to create a linked list from a list of integers
def create_list(arr):
    if not arr:
        return None
    dummy = ListNode(0)
    current = dummy
    for val in arr:
        current.next = ListNode(val)
        current = current.next
    return dummy.next

# Helper function to print a linked list
def print_list(head):
    if not head:
        print("[]")
        return
    
    nodes = []
    current = head
    while current:
        nodes.append(str(current.val))
        current = current.next
    print("[" + ", ".join(nodes) + "]")

if __name__ == "__main__":
    solution = Solution()

    # Test Case 1
    print("Test Case 1:")
    lists1 = [
        create_list([1, 4, 5]),
        create_list([1, 3, 4]),
        create_list([2, 6])
    ]
    print("Input:")
    for l in lists1: print_list(l)
    result1 = solution.mergeKLists(lists1)
    print("Output:", end=" ")
    print_list(result1)
    print()

    # Test Case 2
    print("Test Case 2:")
    lists2 = []
    print("Input: []")
    result2 = solution.mergeKLists(lists2)
    print("Output:", end=" ")
    print_list(result2)
    print()
    
    # Test Case 3
    print("Test Case 3:")
    lists3 = [create_list([])]
    print("Input: [[]]")
    result3 = solution.mergeKLists(lists3)
    print("Output:", end=" ")
    print_list(result3)
    print()

    # Test Case 4
    print("Test Case 4:")
    lists4 = [
        create_list([]),
        create_list([1])
    ]
    print("Input:")
    for l in lists4: print_list(l)
    result4 = solution.mergeKLists(lists4)
    print("Output:", end=" ")
    print_list(result4)
    print()
    
    # Test Case 5
    print("Test Case 5:")
    lists5 = [
        create_list([10, 20]),
        create_list([5, 15]),
        create_list([1, 2, 22])
    ]
    print("Input:")
    for l in lists5: print_list(l)
    result5 = solution.mergeKLists(lists5)
    print("Output:", end=" ")
    print_list(result5)
    print()

File: dataset/CoT/GEMINI_gemini-2.5-pro/Task186.java
import java.util.PriorityQueue;
import java.util.Comparator;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class Task186 {

    /**
     * Merges k sorted linked lists into one single sorted linked list.
     * @param lists An array of sorted linked lists.
     * @return The head of the merged sorted linked list.
     */
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }

        PriorityQueue<ListNode> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a.val));

        for (ListNode list : lists) {
            if (list != null) {
                minHeap.add(list);
            }
        }

        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;

        while (!minHeap.isEmpty()) {
            ListNode node = minHeap.poll();
            tail.next = node;
            tail = tail.next;

            if (node.next != null) {
                minHeap.add(node.next);
            }
        }

        return dummy.next;
    }

    // Helper method to create a linked list from an array of integers
    public static ListNode createList(int[] arr) {
        if (arr == null || arr.length == 0) {
            return null;
        }
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        for (int val : arr) {
            current.next = new ListNode(val);
            current = current.next;
        }
        return dummy.next;
    }

    // Helper method to print a linked list
    public static void printList(ListNode head) {
        if (head == null) {
            System.out.println("[]");
            return;
        }
        System.out.print("[");
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val);
            if (current.next != null) {
                System.out.print(", ");
            }
            current = current.next;
        }
        System.out.println("]");
    }

    public static void main(String[] args) {
        Task186 solution = new Task186();

        // Test Case 1
        System.out.println("Test Case 1:");
        ListNode[] lists1 = {
            createList(new int[]{1, 4, 5}),
            createList(new int[]{1, 3, 4}),
            createList(new int[]{2, 6})
        };
        System.out.print("Input: ");
        for (ListNode l : lists1) printList(l);
        ListNode result1 = solution.mergeKLists(lists1);
        System.out.print("Output: ");
        printList(result1);
        System.out.println();

        // Test Case 2
        System.out.println("Test Case 2:");
        ListNode[] lists2 = {};
        System.out.println("Input: []");
        ListNode result2 = solution.mergeKLists(lists2);
        System.out.print("Output: ");
        printList(result2);
        System.out.println();
        
        // Test Case 3
        System.out.println("Test Case 3:");
        ListNode[] lists3 = {createList(new int[]{})};
        System.out.println("Input: [[]]");
        ListNode result3 = solution.mergeKLists(lists3);
        System.out.print("Output: ");
        printList(result3);
        System.out.println();

        // Test Case 4
        System.out.println("Test Case 4:");
        ListNode[] lists4 = {
            createList(new int[]{}),
            createList(new int[]{1})
        };
        System.out.print("Input: ");
        for (ListNode l : lists4) printList(l);
        ListNode result4 = solution.mergeKLists(lists4);
        System.out.print("Output: ");
        printList(result4);
        System.out.println();
        
        // Test Case 5
        System.out.println("Test Case 5:");
        ListNode[] lists5 = {
            createList(new int[]{10, 20}),
            createList(new int[]{5, 15}),
            createList(new int[]{1, 2, 22})
        };
        System.out.print("Input: ");
        for (ListNode l : lists5) printList(l);
        ListNode result5 = solution.mergeKLists(lists5);
        System.out.print("Output: ");
        printList(result5);
        System.out.println();
    }
}